### PR TITLE
Rework plugin loading

### PR DESCRIPTION
**Pull request by X9VoiD**
## Changes

- Rewritten loading logic from scratch.
- Allow plugins (contained in a subdirectory) to load unmanaged dll dependency of their dependencies.
- Unified loading logic for Daemon and UX.
- Lock-less parallel loading wherever possible.

## Tested

- [x] Plugin loading
- [x] Unmanaged dll loading

### CODE REVIEW

**Comment by X9VoiD** [comment]
Unmanaged dll loading is tested working now.

Screenshot from 2020-11-04 11-20-25

File: OpenTabletDriver/PluginManager.cs [hunk 1]
```diff
@@ -1,105 +1,166 @@
...
         {
-            var children = from type in Types
-                where typeof(T).IsAssignableFrom(type)
-                select type;
-            
-            return new List<TypeInfo>(children);
+            var types = from asm in context.Assemblies
+                        from type in asm.GetExportedTypes()
+                        where type.IsPluginType()
+                        select type;
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:102]: Readd the original try load assembly method here

File: OpenTabletDriver/PluginManager.cs [hunk 2]
```diff
@@ -111,48 +172,26 @@ private static bool ParametersValid(ParameterInfo[] parameters, object[] args)
...
-                _ = asm.DefinedTypes;
-                return true;
-            }
-            catch
+            if (!Silent)
             {
-                return false;
+                lock (logLock)
+                    Plugin.Log.Write("Plugin", msg, level);
             }
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:194]: This shouldn't be needed, Log.Write just invokes an event which shouldn't cause any threading issues.

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

**Comment by X9VoiD** [comment]
Changes pushed.

File: OpenTabletDriver.Daemon/DriverDaemon.cs [hunk 3]
```diff
@@ -278,10 +277,15 @@ private void SetToolSettings()
...
+                {
+                    tool.Initialize();
                     Tools.Add(tool);
-                else
-                    Log.Write("Tool", $"Failed to initialize {plugin.Name} tool.", LogLevel.Error);
+                } 
+                catch (Exception e)
+                {
+                    Log.Write("Tool", $"Failed to initialize {plugin.Name} tool: {e.Message}", LogLevel.Error);
+                }
```
Comment (InfinityGhost) [inline OpenTabletDriver.Daemon/DriverDaemon.cs:288]: Why did this change? If anything we should match it on ITool as well if we are going to keep it.

Comment (X9VoiD) [reply OpenTabletDriver.Daemon/DriverDaemon.cs:288]: Ah, right. Then should we change ITool to return void instead of bool?

Comment (InfinityGhost) [reply OpenTabletDriver.Daemon/DriverDaemon.cs:288]: Might as well, but this isn't a change for this type of PR imo.

Comment (X9VoiD) [reply OpenTabletDriver.Daemon/DriverDaemon.cs:288]: I see, will revert it then for this PR.

Comment (X9VoiD) [reply OpenTabletDriver.Daemon/DriverDaemon.cs]: Revert request is pushed.

File: OpenTabletDriver/PluginManager.cs [hunk 4]
```diff
@@ -1,105 +1,166 @@
...
+
+            // If there are plugins found outside subdirectories then load into FallbackPluginContext
+            // This fallback does not support loading unmanaged dll if the default loader fails
+            // We don't worry with duplicate entries here since CLR won't load duplicate assemblies of the same file
+
+            foreach(var plugin in Directory.EnumerateFiles(AppInfo.Current.PluginDirectory, "*.dll"))
             {
-                return false;
+                var pluginFile = new FileInfo(plugin);
+                Log($"Loading deprecated plugin '{plugin}'");
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:70]: Would prefer to call them something like "global plugins" which don't depend on anything besides the base plugin library.

File: OpenTabletDriver/PluginManager.cs [hunk 5]
```diff
@@ -1,105 +1,166 @@
...
+            types.AsParallel().ForAll(type =>
+            {
+                if (!type.IsPlatformSupported())
+                {
+                    Log($"Plugin '{type.FullName}' incompatible with current OS", LogLevel.Warning);
+                    return;
+                }
+                if (type.IsPluginIgnored())
+                {
+                    Log($"Plugin '{type.FullName}' ignored", LogLevel.Debug);
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:113]: Do not log cases of ignored plugins.

File: OpenTabletDriver/PluginManager.cs [hunk 6]
```diff
@@ -1,105 +1,166 @@
...
+                        where asm.IsLoadable()
+                        from type in asm.GetExportedTypes()
+                        where type.IsPluginType()
+                        select type;
+
+            types.AsParallel().ForAll(type =>
+            {
+                if (!type.IsPlatformSupported())
+                {
+                    Log($"Plugin '{type.FullName}' incompatible with current OS", LogLevel.Warning);
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:108]:
```cs
Log($"Plugin '{type.FullName}' is not supported on {SystemInfo.CurrentPlatform}", LogLevel.Info);
```

File: OpenTabletDriver/PluginManager.cs [hunk 7]
```diff
@@ -1,105 +1,166 @@
...
-                    where TypeImplementsPlugin(type)
-                    select type.GetTypeInfo();
+                try
+                {
+                    var type = PluginTypes.FirstOrDefault(t => t.FullName == name);
+                    var matchingConstructors = from ctor in type?.GetConstructors()
+                                               let parameters = ctor.GetParameters()
+                                               where parameters.Length == args.Length
+                                               where args.IsValidParameterFor(parameters)
+                                               select ctor;
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:141]: 4 spaces for tabbing in LINQ queries, keeps tabbing consistent.

File: OpenTabletDriver/PluginManager.cs [hunk 8]
```diff
@@ -1,105 +1,166 @@
...
+                                from type in asm.DefinedTypes
+                                where type.IsPublic && !(type.IsInterface || type.IsAbstract)
+                                where type.IsPluginType() && type.IsPlatformSupported()
+                                select type;
+
+            internalTypes.AsParallel().ForAll(t => pluginTypes.Add(t.GetTypeInfo()));
+
+            // "Plugins" are directories that contain managed and unmanaged dll
+            //  These dlls are loaded into a PluginContext per directory
+
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:48]: Spacing pls

File: OpenTabletDriver/PluginManager.cs [hunk 9]
```diff
@@ -1,105 +1,166 @@
...
+            foreach(var plugin in Directory.EnumerateFiles(AppInfo.Current.PluginDirectory, "*.dll"))
             {
-                return false;
+                var pluginFile = new FileInfo(plugin);
+                Log($"Loading deprecated plugin '{plugin}'");
+                LoadPlugin(fallbackPluginContext, plugin);
             }
+
+            // Populate PluginTypes so UX and Daemon can access them
+
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:75]: Dood

File: OpenTabletDriver/PluginManager.cs [hunk 10]
```diff
@@ -1,105 +1,166 @@
...
-            var children = from type in Types
-                where typeof(T).IsAssignableFrom(type)
-                select type;
-            
-            return new List<TypeInfo>(children);
+            var types = from asm in context.Assemblies
+                        where asm.IsLoadable()
+                        from type in asm.GetExportedTypes()
+                        where type.IsPluginType()
+                        select type;
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:102]: LINQ query tabbing, 4 spaces

File: OpenTabletDriver/PluginManager.cs [hunk 11]
```diff
@@ -1,105 +1,166 @@
...
-                return AssemblyLoadContext.Default.LoadFromAssemblyPath(path);
-            }
-            catch (BadImageFormatException)
-            {
-                Log.Write("Plugin", $"Failed to initialize {path}, incompatible plugin", LogLevel.Warning);
-                return null;
-            }
+            var children = from type in PluginTypes
+                           where typeof(T).IsAssignableFrom(type)
+                           select type;
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:158]: You probably get the gist lol

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

File: OpenTabletDriver/PluginManager.cs (hunk 4 above)
Comment (X9VoiD) [reply OpenTabletDriver/PluginManager.cs:70]: Will stand-alone plugin work? or independent.

**Comment by X9VoiD** [comment]
Requested changes pushed

File: OpenTabletDriver/PluginManager.cs (hunk 4 above)
Comment (InfinityGhost) [reply OpenTabletDriver/PluginManager.cs]: Standalone implies something different, and independent might be the closest thing to it.

File: OpenTabletDriver/PluginManager.cs [hunk 12]
```diff
@@ -1,105 +1,162 @@
...
-        public static Collection<TypeInfo> Types
+        public static IReadOnlyCollection<TypeInfo> PluginTypes => pluginTypes;
+
+        private static bool Silent;
+        private readonly static PluginContext fallbackPluginContext = new PluginContext();
+        private readonly static ConcurrentBag<PluginContext> plugins = new ConcurrentBag<PluginContext>();
+        private readonly static ConcurrentBag<TypeInfo> pluginTypes = new ConcurrentBag<TypeInfo>();
+        private readonly static IEnumerable<Type> libTypes = from type in Assembly.GetAssembly(typeof(IDriver)).GetExportedTypes()
+                                                             where type.IsAbstract || type.IsInterface
+                                                             select type;
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:26]: Tabbing again lol

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

File: OpenTabletDriver/PluginManager.cs (hunk 12 above)
Comment (X9VoiD) [reply OpenTabletDriver/PluginManager.cs]: Pushed. I made sure that I missed nothing now lol

File: OpenTabletDriver/PluginManager.cs [hunk 13]
```diff
@@ -143,16 +190,15 @@ private static bool CanLoadAssembly(Assembly asm)
...
+        public static void Log(string msg, LogLevel level = LogLevel.Info)
         {
-            return from asm in AppDomain.CurrentDomain.GetAssemblies()
-                where CanLoadAssembly(asm)
-                from type in GetTypes(asm)
-                select type;
-        });
+            if (!Silent)
+                Plugin.Log.Write("Plugin", msg, level);
+        }
```
Comment (InfinityGhost) [inline OpenTabletDriver/PluginManager.cs:202]: I think we could remove the need for this function along with the silent variable by moving the `Log.Message` event hook after plugin loading

**Review by InfinityGhost — APPROVED** [review]

**Pull request merged**